domain/port/state/state.go

```diff
@@ -0,0 +1,269 @@
+				changedEndpoints = append(changedEndpoints, endpoint.UUID)
+				unitPortRanges = append(unitPortRanges, transform.Slice(reconciledOpenPorts, func(p network.PortRange) unitPortRange {
+					return unitPortRange{
+						ProtocolID:       protocolMap[p.Protocol],
+						FromPort:         p.FromPort,
+						ToPort:           p.ToPort,
+						UnitEndpointUUID: endpoint.UUID,
+					}
+				})...)
+			}
+		}
+
+		// Clear out all existing open ports for endpoint undergoing changes.
+		if len(changedEndpoints) > 0 {
+			err = tx.Query(ctx, clearPortRange, changedEndpoints).Run()
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+
+		// Insert the freshly calculated port ranges
+		if len(unitPortRanges) > 0 {
+			err = tx.Query(ctx, insertPortRange, unitPortRanges).Run()
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+
+		return nil
+	})
+}
+
+// ensureEndpoints ensures that the given endpoints are present in the database.
+// Return all endpoints under action with their corresponding UUIDs.
+func (st *State) ensureEndpoints(ctx context.Context, tx *sqlair.TX, unitUUID unitUUID, endpointsUnderAction endpoints) ([]endpoint, error) {
+	getUnitEndpoints, err := st.Prepare(`
+SELECT &endpoint.*
+FROM unit_endpoint
+WHERE unit_uuid = $unitUUID.unit_uuid
+AND endpoint IN ($endpoints[:])
+`, endpoint{}, unitUUID, endpointsUnderAction)
+	if err != nil {
+		return nil, errors.Annotate(err, "preparing get unit endpoints statement")
+	}
+
+	insertUnitEndpoint, err := st.Prepare("INSERT INTO unit_endpoint (*) VALUES ($unitEndpoint.*)", unitEndpoint{})
+	if err != nil {
+		return nil, errors.Annotate(err, "preparing insert unit endpoint statement")
+	}
+
+	var endpoints []endpoint
+	err = tx.Query(ctx, getUnitEndpoints, unitUUID, endpointsUnderAction).GetAll(&endpoints)
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return nil, errors.Trace(err)
+	}
+
+	foundEndpoints := set.NewStrings()
+	for _, ep := range endpoints {
+		foundEndpoints.Add(ep.Endpoint)
+	}
+
+	// Insert any new endpoints that are required.
+	requiredEndpoints := set.NewStrings(endpointsUnderAction...).Difference(foundEndpoints)
+	newUnitEndpoints := make([]unitEndpoint, requiredEndpoints.Size())
+	for i, requiredEndpoint := range requiredEndpoints.Values() {
+		uuid, err := uuid.NewUUID()
+		if err != nil {
+			return nil, errors.Annotatef(err, "generating UUID for unit endpoint")
+		}
+		newUnitEndpoints[i] = unitEndpoint{
+			UUID:     uuid.String(),
+			UnitUUID: unitUUID.UUID,
+			Endpoint: requiredEndpoint,
+		}
+		endpoints = append(endpoints, endpoint{
+			Endpoint: requiredEndpoint,
+			UUID:     uuid.String(),
+		})
+	}
+
+	if len(newUnitEndpoints) > 0 {
+		err = tx.Query(ctx, insertUnitEndpoint, newUnitEndpoints).Run()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	return endpoints, nil
+}
+
+func (st *State) getProtocolMap(ctx context.Context, tx *sqlair.TX) (map[string]int, error) {
```

Comment:
You don't need this function, because you already selected and populated the IDs and the values in the query to retrieve them, right?

Reply:
Unfortunately I think we still do need this.

Think about the case of a unit where all ports are closed, and we wish to open `443/tcp`.

We need to insert a row into the db for the opened port range, including the id of `tcp`. How do we get the id of `tcp`?